sw/device/lib/crypto/impl/rsa.c

```diff
@@ -63,116 +92,276 @@ static status_t rsa_mode_check(const key_mode_t mode) {
-    const rsa_private_key_t *private_key) {
-  // Check for NULL pointers.
-  if (private_key == NULL || private_key->n.key == NULL ||
-      private_key->d.keyblob == NULL) {
+    const rsa_size_t size, const crypto_blinded_key_t *private_key) {
+  // Check that the key mode is a valid RSA mode.
+  HARDENED_TRY(rsa_mode_check(private_key->config.key_mode));
+
+  // Sideloaded keys are not supported for RSA.
+  if (private_key->config.hw_backed != kHardenedBoolFalse) {
     return OTCRYPTO_BAD_ARGS;
   }
 
-  // Check that the modulus key mode is a valid RSA mode.
-  HARDENED_TRY(rsa_mode_check(private_key->n.key_mode));
-
-  // Check that the key modes match each other.
-  if (private_key->n.key_mode != private_key->d.config.key_mode) {
+  // Check the lengths against the RSA size.
+  size_t key_length;
+  size_t keyblob_length;
+  HARDENED_TRY(
+      otcrypto_rsa_private_key_length(size, &key_length, &keyblob_length));
+  if (private_key->config.key_length != key_length ||
+      private_key->keyblob_length != keyblob_length) {
     return OTCRYPTO_BAD_ARGS;
   }
 
-  // Sideloaded keys are not supported for RSA.
-  if (private_key->d.config.hw_backed != kHardenedBoolFalse) {
+  return OTCRYPTO_OK;
+}
+
+crypto_status_t otcrypto_rsa_private_key_from_exponents(
+    rsa_size_t size, crypto_const_word32_buf_t modulus, uint32_t e,
+    crypto_const_word32_buf_t d_share0, crypto_const_word32_buf_t d_share1,
+    crypto_blinded_key_t *private_key) {
+  if (modulus.data == NULL || d_share0.data == NULL || d_share1.data == NULL ||
+      private_key == NULL || private_key->keyblob == NULL) {
     return OTCRYPTO_BAD_ARGS;
   }
+  HARDENED_TRY(rsa_mode_check(private_key->config.key_mode));
 
-  // Check that the unblinded size for the private exponent matches the length
+  // Ensure that the length of the private exponent shares matches the length
   // of the modulus.
-  if (private_key->d.config.key_length != private_key->n.key_length) {
+  if (d_share0.len != modulus.len || d_share1.len != modulus.len) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the public exponent is odd and greater than 2^16 (see FIPS 186-5,
+  // section A.1.1).
+  if ((e & 1) != 1 || e >> 16 == 0) {
     return OTCRYPTO_BAD_ARGS;
   }
 
+  // Check the mode and lengths for the private key.
+  HARDENED_TRY(private_key_structural_check(size, private_key));
+
+  switch (size) {
+    case kRsaSize2048: {
+      if (private_key->keyblob_length != sizeof(rsa_2048_private_key_t) ||
+          modulus.len != kRsa2048NumWords) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      rsa_2048_private_key_t *sk =
+          (rsa_2048_private_key_t *)private_key->keyblob;
+      hardened_memcpy(sk->n.data, modulus.data, modulus.len);
+      hardened_memcpy(sk->d.data, d_share0.data, d_share0.len);
+      // TODO: RSA keys are currently unblinded, so combine the shares.
+      for (size_t i = 0; i < d_share1.len; i++) {
+        sk->d.data[i] ^= d_share1.data[i];
+      }
```

Comment:
same here, redundant code between case blocks?

Reply:
This is again because the key type is different for each RSA size; maybe the type names should be a little less verbose so that's more visible.